condor_git_config.py

```diff
@@ -91,127 +98,174 @@ class ConfigCache(object):
+        return self.abspath("repo", *rel_paths)
 
     def __enter__(self):
         self._cache_lock.acquire()
+        self._config_meta = self._read_meta()
+        self._refresh()
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        self._write_meta()
+        self._config_meta = None
         self._cache_lock.release()
         return False
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[Path]:
+        assert self._config_meta is not None
         # avoid duplicates from links
         seen = set()
         repo_path = self.repo_path()
         for dir_path, dir_names, file_names in os.walk(repo_path):
-            if '.git' in dir_names:
-                dir_names.remove('.git')
-            dir_names[:] = sorted(dir_names)
+            try:
+                dir_names.remove(".git")
+            except ValueError:
+                pass
+            dir_names.sort()
+            dir_path = Path(dir_path)
             for file_name in sorted(file_names):
-                rel_path = os.path.normpath(os.path.relpath(os.path.join(dir_path, file_name), repo_path))
+                rel_path = (dir_path / file_name).relative_to(repo_path)
                 if rel_path in seen:
                     continue
                 seen.add(rel_path)
                 yield rel_path
 
-    @property
-    def outdated(self):
-        try:
-            with open(self._meta_file, 'r') as raw_meta:
-                meta_data = json.load(raw_meta)
-        except FileNotFoundError:
-            return True
-        else:
-            if meta_data['git_uri'] != self.git_uri or meta_data['branch'] != self.branch:
-                LOGGER.critical('cache %r corrupted by other hook: %r', self, meta_data)
-                raise RuntimeError('config cache %r used for conflicting hooks' % self.cache_path)
-            else:
-                return meta_data['timestamp'] + self.max_age <= time.time()
-
-    def _update_metadata(self):
-        with open(self._meta_file, 'w') as raw_meta:
-            json.dump({'git_uri': self.git_uri, 'branch': self.branch, 'timestamp': time.time()}, raw_meta)
-
-    def refresh(self):
-        if not self.outdated:
+    def _refresh(self):
+        assert self._config_meta is not None
+        self._config_meta["reads"] += 1
+        if self._config_meta["timestamp"] + self.max_age > time.time():
             return
         repo_path = self.repo_path()
-        if not os.path.exists(os.path.join(repo_path, '.git')):
+        self._config_meta["pulls"] += 1
+        if not self.repo_path(".git").exists():
+            branch = [] if not self.branch else ["--branch", self.branch]
             subprocess.check_output(
-                ['git', 'clone', '--quiet', '--branch', str(self.branch), str(self.git_uri), repo_path],
+                ["git", "clone", "--quiet", *branch, self.git_uri, str(repo_path)],
                 timeout=30,
                 universal_newlines=True,
             )
```

Comment:
It is on purpose to not have a try-except here, correct? You want the process to crash?

Reply:
Yes, since this is the first checkout the script cannot recover if this fails. Should I document this with a comment?

Reply:
I think you can leave it as is. I expected this kind of answer, just wanted to ensure it with you.